### DISCUSSION TITLE
EVP_aes_256_gcm mode holding IV -16 bytes space holding garbage values.

### DISCUSSION THREAD

**Discussion by mahenderkk08** [openssl/openssl#24708] [Q&A] [answered]
Hi Team,

We are currently using OpenSSL version 3.1.2 and the FIPS module version 3.0.8 in our setup. For encryption, we are using the API EVP_aes_256_gcm. During our application testing with the gcm mode, the IV value somehow contains 16 bytes of garbage values, so we are now taking 17 bytes and filling the last position with a null terminator value.

Additionally, we have noticed a security vulnerability at https://nvd.nist.gov/vuln/detail/CVE-2023-5363. Is this similar to the issue we have currently noticed, and has it been fixed in version 3.1.4? 

Also sharing below our code snip:

```
EncryptData(const unsigned char* key_data, int key_length, const unsigned char* in, unsigned long size, unsigned char* out, const EVP_CIPHER *EncryptionType, TCHAR* errorMsg, BYTE* IVFromDb, int *nCipherTextLength)
{

	if (!size || size == 0)
	{
		return E_FAIL;
	}

	int cipher_len = 0, len = 0;
	EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
	BYTE *iv = NULL;
	try
	{
		if (!ctx)
		{
			EVP_CIPHER_CTX_cleanup(ctx);
			return E_FAIL;
		}

		**if (EncryptionType == EVP_aes_256_gcm())
		{
			OpenSSL_add_all_ciphers();

			if (IVFromDb != NULL && (strlen((char*)IVFromDb)) > 0)
			{
				iv = IVFromDb;
				iv[16] = '\0';
				SetIVValueToRegistry(iv, 16);
			}
			else
			{
				iv = GetIVValueFromRegistry();
				if (iv == NULL)
				{
					// somehow 16 bytes space holding garbage values, so that here we are taking 17 and null terminator value is filled in the the last position
					iv = new BYTE[17];
					RAND_bytes(iv, 16);
					iv[16] = '\0';
					SetIVValueToRegistry(iv, 16);
				}
				else
				{

				}
			}**

			EVP_CIPHER_CTX* e_ctx = EVP_CIPHER_CTX_new();
			if (!EVP_EncryptInit_ex2(e_ctx, EncryptionType, (const unsigned char*)key_data, iv, NULL))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_EncryptInit_ex2() failed");
				EVP_CIPHER_CTX_free(e_ctx);
				return E_FAIL;
			}

			if (!EVP_EncryptUpdate(e_ctx, out, &len, (const unsigned char*)in, size))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_EncryptUpdate() failed");
				EVP_CIPHER_CTX_free(e_ctx);
				return E_FAIL;
			}
			cipher_len += len;
			if (!EVP_EncryptFinal_ex(e_ctx, out + len, &len))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_EncryptFinal_ex() failed");
				EVP_CIPHER_CTX_free(e_ctx);
				return E_FAIL;
			}
			cipher_len += len;

			if (!EVP_CIPHER_CTX_ctrl(e_ctx, EVP_CTRL_GCM_GET_TAG, TAG_LENGTH, out + cipher_len))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_CIPHER_CTX_ctrl() failed");
				EVP_CIPHER_CTX_free(e_ctx);
				return E_FAIL;
			}
			cipher_len += TAG_LENGTH;

			memset(out + cipher_len, '\0', (size + TAG_LENGTH) - cipher_len);

			*nCipherTextLength = cipher_len;

			EVP_CIPHER_CTX_free(e_ctx);		

		}
		
	}
	catch (...)
	{
		int err = ERR_get_error();
		char buf[256];
		ERR_error_string_n(err, buf, sizeof(buf));	
		EVP_CIPHER_CTX_cleanup(ctx);

		return E_FAIL;
	}
	return S_OK;
}
DecryptData(const unsigned char* key_data, int key_length, const unsigned char* in, unsigned long size, unsigned char* out, unsigned long* ret_size, const EVP_CIPHER *DecryptionType, TCHAR* errorMsg, BYTE* IV)
{
	if ((!size) || size == 0)
	{
		return E_FAIL;
	}
	unsigned char decm[4096] = { 0 };
	int declen = 0, declen2 = 0;
	int text_len = 0;
	int len = 0;
	EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
	BYTE *iv = NULL;
	iv = (IV != NULL) ? IV : NULL;
	try
	{
		if (!ctx)
		{
			EVP_CIPHER_CTX_cleanup(ctx);
			if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"EVP_CIPHER_CTX_new() failed");
			return E_FAIL;
		}

		if (DecryptionType == EVP_aes_256_gcm())
		{
			if (iv == NULL)
			{

				iv = GetIVValueFromRegistry();
				if (iv == NULL)
				{
					return E_FAIL;
				}
			}
			else
			{
			}

			EVP_CIPHER_CTX *d_ctx = EVP_CIPHER_CTX_new();
			if (!EVP_DecryptInit_ex2(d_ctx, DecryptionType, key_data, iv, NULL))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_DecryptInit_ex2() failed");
				return E_FAIL;
			}
			if (!EVP_CIPHER_CTX_ctrl(d_ctx, EVP_CTRL_GCM_SET_TAG, TAG_LENGTH, (unsigned char*)in + (size - TAG_LENGTH)))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_CIPHER_CTX_ctrl() failed");
				EVP_CIPHER_CTX_free(d_ctx);
				return E_FAIL;
			}

			if (!EVP_DecryptUpdate(d_ctx, out, &len, in, size - TAG_LENGTH))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_DecryptUpdate() failed");
				EVP_CIPHER_CTX_free(d_ctx);
				return E_FAIL;
			}
			text_len += len;

			if (!EVP_DecryptFinal_ex(d_ctx, out + len, &len))
			{
				if (errorMsg != NULL) _tcscpy_s(errorMsg, MAX_PATH, L"GCM : EVP_DecryptFinal_ex() failed");
				EVP_CIPHER_CTX_free(d_ctx);
				return E_FAIL;
			}
			text_len += len;

			memset(out + text_len, '\0', size - text_len);

			*ret_size = text_len;
			EVP_CIPHER_CTX_free(d_ctx);
		}
		if (ctx != NULL)
			EVP_CIPHER_CTX_cleanup(ctx);
	}
	catch (...)
	{
		if (ctx != NULL)
			EVP_CIPHER_CTX_cleanup(ctx);
		return E_FAIL;
	}
	return S_OK;
}
```

**Comment by t8m** [comment, marked as answer]
The IV value should never be _null terminated_ as the value is binary. It is not a C-string.

The default IV length for AES-GCM is 12 bytes. You're not overriding it anywhere. That means anything after these 12 bytes in your IV buffer is ignored.

CVE-2023-5363 is only remotely related to your code and not really relevant here.

_ During our application testing with the gcm mode, the IV value somehow contains 16 bytes of garbage values, so we are now taking 17 bytes and filling the last position with a null terminator value._

This does not make much sense to me. The IV value must be coming from your application so if there is a garbage, the application must have written garbage there or kept it uninitialized.

**Reply by paulidale** [reply]
The FIPS provider creates a random IV internally, it's not possible to push one in from outside the FIPS boundary.
For decryption, the IV is used.

**Reply by t8m** [reply]
Ah, right. But the application code above does not use the internal IV generation CTRLs.

**Reply by paulidale** [reply]
Perhaps it's not using the FIPS provider's implementation.  `EVP_aes_256_gcm()` doesn't set a property query.

I'd suggest using `EVP_CIPHER_fetch()` to ensure the FIPS version is used.

**Reply by mahenderkk08** [reply]
Hi @paulidale , EVP_CIPHER_fetch() not sure. Could you please provide a sample example that uses it? Just a reference will help to test my application's functionality. Thanks in advance.

**Reply by paulidale** [reply]
Look in the [demo/cipher/aesgcm.c](https://github.com/openssl/openssl/blob/master/demos/cipher/aesgcm.c) file in the sources you downloaded:

**Comment by paulidale** [comment]
One other thought here.  Why are you storing the IV in the registry?  That is a **very bad** place to put it.

When using GCM mode it is **critical** that a different IV be used whenever you encrypt with the same key.  Repeating an IV, even once, will break the encryption.

Having the IV in the registry seems like it would be extremely prone to reuse problems.  The IV should be stored along with the encrypted data.

**Reply by mahenderkk08** [reply]
Hi
Our application will work with multiple servers. We have a dependency with the same key. Whenever we join the same organization, we use the key to allow the user to join the new server. We check validation with the same encryption data by storing it.

**Reply by paulidale** [reply]
Then you **must** ensure that every IV **is different**.  Without that the encryption will be **broken**.
Storing the IV in the registry is definitely a poor choice here.  Transmit the IV along with the encrypted data and don't store it anywhere else.

**Reply by mahenderkk08** [reply]
In our case,  to join the old organization our application server password should match here we need the same encryption check.  so we are keeping the IV in the registry securely for reuse.

**Reply by paulidale** [reply]
You might as well **not encrypt the data** if you reuse the IV.  It will be **exactly as secure** as what you are doing and much faster.  You **will** leak the plaintext and **likely compromise** your encryption key.

The first answer to [this question](https://crypto.stackexchange.com/questions/26790/how-bad-it-is-using-the-same-iv-twice-with-aes-gcm) is telling.  Search for "gcm IV reuse" for many more including detailed explanations.

**Comment by paulidale** [comment]
Setting the IV like you are doing will result in your usage being non compliant to the FIPS standards.  FIPS 140 mandates that the GCM IV must be generated inside the module.  The correct approach is to not set the IV when encrypting and to retrieve and save it for decryption using `EVP_CIPHER_CTX_get_params()`.

**Reply by t8m** [reply]
I'm also getting an impression that they use AES-GCM for something that should actually use some completely different algorithm, maybe from a completely different family even than AEAD cipher.
